numpyro/util.py

```diff
@@ -221,7 +286,9 @@ def _body_fn(i, vals):
 
     collection = jnp.zeros((collection_size,) + init_val_flat.shape)
     if not progbar:
```

Comment:
With `num_chains=1` and `progress_bar=False`, I guess we want to use the old `_body_fn`. You can add an option to `progbar_opts` to control the behavior I guess, e.g. adding `num_chains` to [progbar_opts](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/mcmc.py#L354), removing [this branch](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/mcmc.py#L264), and revising the logic here into
```python
if not progbar:
    last_val, collection, _, _ = fori_loop(0, upper, _body_fn, (init_val, collection, start_idx, thinning))
elif num_chains > 1:
    progress_bar_fori_loop = progress_bar_factory(upper)
    _body_fn_pbar = progress_bar_fori_loop(_body_fn)
    last_val, collection, _, _ = fori_loop(0, upper, _body_fn_pbar, (init_val, collection, start_idx, thinning))
```

Reply:
Sorry for the delay!

I don't understand what the API should be for the new progress bar.

Should `progress_bar` in `MCMC` still control the old progress bar (for un-compiled loops)? You had suggested in an earlier comment that we could add `host_callback=False` in the MCMC class and pass that to `fori_collect` so that people can test it at first. 

I don’t quite get how `num_chains` should control the behaviour of the compiled progress bar. Do you mean that the progress bar for complied sampling should be on if `num_chains>1`?

Reply:
Currently, with `num_chains > 1` and `chain_method="parallel"`, we [switch progress_bar to False](https://github.com/pyro-ppl/numpyro/blob/ce6cac87ec715ea467a353160dcb56a1827b28a9/numpyro/infer/mcmc.py#L266) because we don't have support for it yet. And IIUC, your PR tries to address that issue. Back then, I thought you are going to do host callback for every MCMC step, so I wondered about the overhead. That was the reason that I suggested having a new argument to test things out first. For now, with your strategy of updating progressing only after each 5% of the total progress, I think the overhead of using host_callback would be pretty small. So you can go ahead and make it work for the default behavior of `num_chains > 1, progress_bar=True, chain_method="parallel"`.

Reply:
Unfortunately, I’m still confused about what the API should be.

To summarise, in this PR we currently have the following behaviour:

• `num_chains=1`, `progress_bar=False` -> compiled sampler, and uses the new progress bar 
• `num_chains=1`, `progress_bar=True` -> “python loop” sampler, and uses the old progress bar 
• `num_chains>1`, `progress_bar=False` -> compiled sampler, and uses the new progress bar. 
• `num_chains>1`, `progress_bar=True` -> compiled sampler, and uses the new progress bar. This has the same behaviour as above because of that [line](https://github.com/pyro-ppl/numpyro/blob/ce6cac87ec715ea467a353160dcb56a1827b28a9/numpyro/infer/mcmc.py#L266) in`run()`

So my questions are:
1. Which of these behaviours should change? 
2. It might be good to be able to have the compiled sampler without the progress bar (for both 1 or several chains). How should we do this? The problem is that `progress_bar=True/False` controls that behaviour. Ideally there would be another argument `compiled=True/False` that controls whether it’s compiled. Though I get that changing the API might not be a good idea.

Reply:
> num_chains=1, progress_bar=False -> compiled sampler, and uses the new progress bar
> num_chains>1, progress_bar=False -> compiled sampler, and uses the new progress bar.

I think those are unexpected.

> It might be good to be able to have the compiled sampler without the progress bar

I think if users want progress bars, they can set `progress_bar=True`. Otherwise, they can set `progress_bar=False`. We can have a new argument for users to test the case `num_chains>1, progress_bar=False` but as I mentioned above, it seems to me that the overhead is likely small, so we might not need to introduce that new argument.

Reply:
To be clear on what definitions I using:
- _“compiled”_: the sampler that uses `fori_loop(body_fn)` 
- _“non-compiled”_: the sampler that uses `python_loop(jit(body_fn))`
- _“the new progress bar”_: the tqdm progress bar that is added in this PR. This runs for the compiled sampler
- _“the old progress bar”_ the tqdm progress bar that is in the current master branch of numpyro. This runs for the un-compiled sampler.

----

> I think those are unexpected.

Which bit specifically is unexpected, and what should then be the correct behaviour? 

> Regarding compiled sampler, I guess you meant to display the progress bar even when users set `progress_bar=False`

This is what is currently happening on this branch. See below for my issues with the argument `progress_bar`


> We can have a new argument for users to test the case num_chains>1, progress_bar=False but as I mentioned above, it seems to me that the overhead is likely small, so we might not need to introduce that new argument. What do you think?

I wasn't actually worried about the overhead as I agree that it's very small. I was more thinking about turning off the progress bar for compiled samplers in general (See addition _A1_ below)


> Regarding compiled sampler, I guess you meant to display the progress bar even when users set progress_bar=False, i.e. compiled means fori_loop(body_fn) vs non-compiled means python_loop(jit(body_fn))?

That's indeed what I mean by compiled vs non-compiled (see definitions above). I do think we should have the progress bar for a compiled sampler with 1 chain. I thought that was always the point of this PR. Though from re-reading the original [issue](https://github.com/pyro-ppl/numpyro/issues/309), I realised it was originally about `num_chains>1` (this is probably why I was getting confused :) ). Do you think that the compiled sampler for 1 chain should _not_ display a progress bar?

----

The current behaviour and API of this branch is given by the 4 cases in my [previous comment](https://github.com/pyro-ppl/numpyro/pull/900#discussion_r585462473). I think this behaviour seems good. Note that all samplers (compiled, non-compiled, 1 chain, many chains) have a progress bar.

My thought was that it might be good to have an argument in `MCMC` allowing you to run the compiled sampler without the progress bar (let’s call this addition _A1_). This would apply for both single chain and many chain samplers. This is for when you run MCMC within a larger workflow and want it to be fast but don’t want to see the progress bar. However I get that this is a specific use-case. So perhaps it’s not necessary!

In any case (ie: whether or not we add addition _A1_), the argument `progress_bar` now controls whether or not the sampler is compiled, and not the progress bar. This is the main reason why I suggested added `compiled=True/False` to control whether or not the sampler is compiled. 
However I see your point that `single_chain_compiled` might be better as only a single chain can be non-compiled. And I agree that this might be a confusing message to users!

Reply:
> Which bit specifically is unexpected, and what should then be the correct behaviour?

I think the correct behavior is to not have progress bar when setting `progress_bar=False` and having progress bar when setting `progress_bar=True`. Currently, NumPyro does not have progress bar for multiple chains and that is also unexpected. That is why issue #309 is still open for a long time.

> Do you think that the compiled sampler for 1 chain should not display a progress bar?

It depends on the API I guess. With the current API, the compiled sampler for 1 chain (equivalently, progress_bar=False, num_chains=1) should not display a progress bar because `progress_bar=False`. Let's discuss having a different API in another thread. What do you think?

It seems to me that we agree on having progress bar on num_chains>1 is nice. How about creating a new issue/feature request for always displaying progress bar, so we can address it in another PR? I think we need to discuss first if it is necessary to introduce a new argument for that feature. Some specific usage cases, like when MCMC is very fast to sample, a lot of samples are required while users only have 1 device to run a single chain, and users want to see the 5%-bin progress, might be important. I guess the dispatching cost of the current `progress_bar=True` approach (for single chain, in master branch) will be large in that case but it is not currently clear to me why we need to display progress bar if it is so fast to sample, so would like to have more discussion for that usage. There might be other usage cases that I overlooked, so having a separate thread would be very helpful too. :)

Reply:
Ah I think I understand now, thanks for explaining!

So I'll make some fixes to make sure the `progress_bar=True/False` maps to whether there is a progress bar. 

And I agree that a separate PR would be good to discuss whether to always show the progress bar (ex: for 1 compiled chain)

Reply:
Thank you, @jeremiecoullon! Sorry for making confusion. I'm happy that the path ahead is clear now. :)

Reply:
Thanks, this looks correct to me.